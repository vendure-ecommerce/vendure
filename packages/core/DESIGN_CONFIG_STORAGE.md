Design doc created after some back and forth with Claude Code.

All the details are not 100%, but the general direction is there as a good basis:

## Core Components

### Settings Store Field definition

```ts
// Simplified field configuration for MVP
interface SettingsStoreFieldConfig {
    name: string;
    scope?: SettingsStoreScopeFunction; // Defaults to SettingsStoreScopes.global
    public?: boolean;      // Available via Shop API
    readonly?: boolean;    // Cannot be modified via GraphQL API
    requiresPermission?: Array<Permission | string> | Permission | string;
    validate?: (
        value: any,
        injector: Injector,
        ctx: RequestContext,
    ) => string | LocalizedString[] | void | Promise<string | LocalizedString[] | void>;
}

// Registration interface for plugins
interface SettingsStoreRegistration {
    namespace: string; // e.g., 'dashboard', 'payment', 'shipping'
    fields: SettingsStoreFieldConfig[];
}

// Scope function type - returns string to append to key for scoping
// E.g. to store settings for the active user
type SettingsStoreScopeFunction = (params: {
    key: string;
    value?: any;
    ctx: RequestContext;
}) => string;


// Pre-made scope functions
export const SettingsStoreScopes = {
    global: (): string => '',
    user: ({ ctx }: { ctx: RequestContext }): string => ctx.activeUserId || '',
    channel: ({ ctx }: { ctx: RequestContext }): string => ctx.channelId || '',
    userAndChannel: ({ ctx }: { ctx: RequestContext }): string =>
        `${ctx.activeUserId || ''}:${ctx.channelId || ''}`
};
```

### Storage

Single TypeORM entity handles all storage needs:

**Benefits:**

- âœ… Single table design - simple and efficient
- âœ… Built-in TypeORM integration with existing Vendure patterns
- âœ… Automatic unique constraint handling for scoped keys
- âœ… JSON serialization handled transparently
- âœ… Standard Vendure entity lifecycle (timestamps, etc.)
-

```typescript
// TypeORM Entity for storage
@Entity()
export class SettingsStoreEntry extends VendureEntity {

    @Index()
    @Column()
    key: string;

    @Column('json', { nullable: true })
    value: JSONCompatible<any> | null; // JSON-serialized value

    @Index()
    @Column({ nullable: true })
    scopeKey: string | null; // Result of scope function

    // Unique constraint for key + scopeKey combination
@Index('key_scope_unique', ['key', 'scopeKey'], { unique: true })
}
```

## ðŸ”§ **API Design**

### Service Interface

Simplified service interface for MVP:

```typescript

@Injectable()
export class SettingsStoreService {
    /**
     * Register key-value fields (typically called during VendureConfig processing)
     */
    register(registration: SettingsStoreRegistration): void;

    /**
     * Get value with automatic scoping and permission check
     * Returns undefined if key doesn't exist or user lacks permission
     */
    async get<T = any>(key: string, ctx: RequestContext): Promise<T | undefined>;

    /**
     * Set value with automatic scoping and permission check
     * Value must be JSON-serializable
     */
    async set<T = any>(key: string, value: T, ctx: RequestContext): Promise<void>;

    /**
     * Get multiple values efficiently
     */
    async getMany(keys: string[], ctx: RequestContext): Promise<Record<string, any>>;

    /**
     * Set multiple values in a transaction
     */
    async setMany(values: Record<string, any>, ctx: RequestContext): Promise<void>;

    /**
     * Get field definition for a key
     */
    getFieldDefinition(key: string): SettingsStoreFieldConfig | undefined;

    /**
     * Validate a value against its field definition
     */
    async validateValue(
        key: string,
        value: any,
        ctx: RequestContext
    ): Promise<string | LocalizedString[] | void>;
}
```

### Plugin Integration

Plugin registration via VendureConfig:

```typescript
// Dashboard plugin registration for user-specific UI settings
@VendurePlugin({
    // ... other plugin config
    configuration: (config: VendureConfig) => {
        // Register key-value fields via main VendureConfig
        config.settingsStoreFields = {
            ...config.settingsStoreFields,
            dashboard: [
                {
                    name: 'theme',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated],
                    validate: (value) => {
                        if (!['light', 'dark', 'auto'].includes(value)) {
                            return 'Theme must be light, dark, or auto';
                        }
                    }
                },
                {
                    name: 'tableFilters',
                    scope: SettingsStoreScopes.userAndChannel, // User-specific per channel
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                },
                {
                    name: 'visibleColumns',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                },
                {
                    name: 'notifications',
                    scope: SettingsStoreScopes.user,
                    public: true,
                    requiresPermission: [Permission.Authenticated]
                }
            ]
        };
        return config;
    }
})
export class DashboardPlugin {
    // Plugin implementation...
}
```

### GraphQL Schema

Simplified generic GraphQL schema:

```graphql
# Simplified field definition for MVP
type SettingsStoreFieldDefinition {
    name: String!
    public: Boolean!
    readonly: Boolean!
    requiresPermission: [Permission!]
}

# Generic input/output types using JSON
input SettingsStoreInput {
    key: String!
    value: JSON!
}

# Generic queries and mutations
extend type Query {
    """
    Get value for a specific key (automatically scoped)
    """
    getSettingsStoreValue(key: String!): JSON

    """
    Get multiple settings store pairs (each automatically scoped)
    """
    getSettingsStoreValues(keys: [String!]!): JSON
}

extend type Mutation {
    """
    Set a single settings store pair (automatically scoped)
    """
    setSettingsStoreValue(input: SettingsStoreInput!): Boolean!

    """
    Set multiple settings store pairs in a transaction (each automatically scoped)
    """
    setSettingsStoreValues(inputs: [SettingsStoreInput!]!): Boolean!
}
```

### Plugin Usage with Generic GraphQL API

Any plugin can use these generic operations:

```typescript
// Save user preference - automatically scoped to user
const SAVE_THEME = gql`
    mutation SaveTheme($key: String!, $value: JSON!) {
        setSettingsStoreValue(input: { key: $key, value: $value })
    }
`;

await adminClient.query(SAVE_THEME, {
    variables: { key: 'dashboard.theme', value: 'dark' }
});

// Get multiple settings efficiently
const GET_SETTINGS = gql`
    query GetSettings($keys: [String!]!) {
        getSettingsStoreValues(keys: $keys)
    }
`;

const result = await adminClient.query(GET_SETTINGS, {
    variables: { keys: ['dashboard.theme', 'dashboard.tableFilters'] }
});

// Result: { dashboard.theme: 'dark', dashboard.tableFilters: {...} }
```

## ðŸ”’ **Security & Permissions**

### Permission Model

Simplified permission checking for MVP:

```typescript
// Permission checking based on field configuration (similar to custom fields)
export function userHasSettingsStorePermission(
    ctx: RequestContext,
    fieldDef: SettingsStoreFieldConfig
): boolean {
    // Shop API: only public fields
    if (ctx.apiType === 'shop' && fieldDef.public === true) {
        return true;
    }

    // Admin API: check required permissions
    const requiredPermissions = fieldDef.requiresPermission ?? [];
    if (requiredPermissions.length > 0) {
        return ctx.userHasPermissions(requiredPermissions);
    }

    // For fields without explicit permissions, require basic authentication
    // More granular permission logic can be implemented based on scope function behavior
    return ctx.isAuthorized();
}
```

## ðŸš€ **Implementation Plan**

### MVP (v3.4.0) - Core Settings Store

- [x] Create `SettingsStoreEntry` TypeORM entity with `scope` column
- [x] Implement `SettingsStoreService` with basic CRUD operations
- [x] Add flexible scoping logic using scope functions
- [x] Implement pre-made scope functions (`global`, `user`, `channel`, `userAndChannel`)
- [x] Add permission checking with `requiresPermission` support
- [x] Create GraphQL schema and resolvers
- [x] Add `settingsStoreFields` property to VendureConfig
- [x] Process `settingsStoreFields` during config initialization
- [x] JSON serialization/deserialization
- [x] Basic validation support

### Future Enhancements

- [ ] Type safe inputs and return types, like with custom fields
- [ ] Admin UI components
- [ ] Advanced validation and type safety

## ðŸŽ¨ **Usage Examples**

### User-Specific Dashboard Settings

```typescript
// 1. Register dashboard fields via VendureConfig
const config: VendureConfig = {
    // ... other config
    settingsStoreFields: {
        dashboard: [
            {
                name: 'theme',
                scope: SettingsStoreScopes.user,
                public: true,
                requiresPermission: [Permission.Authenticated],
                validate: (value) => {
                    if (!['light', 'dark', 'auto'].includes(value)) {
                        return 'Theme must be light, dark, or auto';
                    }
                }
            }
        ]
    }
};

// 2. Save user preference - automatically scoped to current user
await settingsStoreService.set('dashboard.theme', 'dark', ctx);

// 3. Retrieve on dashboard load - automatically scoped to current user  
const userTheme = await settingsStoreService.get('dashboard.theme', ctx);

// 4. Update via generic GraphQL mutation
mutation
{
    setSettingsStoreValue(input
:
    {
        key: "dashboard.theme"
        value: "dark"
    }
)
}
```

### Complex Data Storage

```typescript
// Save complex JSON data - all JSON-serializable
const tableFilters = {
    productList: {
        categories: ['electronics'],
        priceRange: { min: 100, max: 500 },
        inStock: true
    },
    orderList: {
        status: ['PaymentSettled', 'PartiallyFulfilled'],
        dateRange: { start: '2024-01-01', end: '2024-12-31' }
    }
};

await settingsStoreService.set('dashboard.tableFilters', tableFilters, ctx);

// Retrieve complex data
const savedFilters = await settingsStoreService.get<typeof tableFilters>('dashboard.tableFilters', ctx);
```

### Mixed Scoping Example

```typescript
// Register fields with different scopes via VendureConfig
const config: VendureConfig = {
    settingsStoreFields: {
        app: [
            {
                name: 'userTheme',
                scope: SettingsStoreScopes.user,
                public: true,
                requiresPermission: [Permission.Authenticated]
            },
            {
                name: 'channelCurrency',
                scope: SettingsStoreScopes.channel,
                public: false,
                requiresPermission: [Permission.ReadSettings]
            },
            {
                name: 'maintenanceMode',
                scope: SettingsStoreScopes.global,
                public: false,
                requiresPermission: [Permission.SuperAdmin]
            }
        ]
    }
};

// Each key is automatically scoped when accessed
const userTheme = await settingsStoreService.get('app.userTheme', ctx);         // User-scoped
const currency = await settingsStoreService.get('app.channelCurrency', ctx);    // Channel-scoped  
const maintenance = await settingsStoreService.get('app.maintenanceMode', ctx); // Global
```

### Advanced Scoping Examples

```typescript
// User AND channel specific settings (same user, different channels)
const config: VendureConfig = {
    settingsStoreFields: {
        dashboard: [
            {
                name: 'tableFilters',
                scope: SettingsStoreScopes.userAndChannel, // User-specific per channel
                public: true,
                requiresPermission: [Permission.Authenticated]
            }
        ]
    }
};

// Custom scope function example
const customDateBasedScope: SettingsStoreScopeFunction = ({ ctx }) => {
    const now = new Date();
    const quarter = Math.floor(now.getMonth() / 3) + 1;
    return `${ctx.activeUserId || ''}:${now.getFullYear()}Q${quarter}`;
};

// Register field with custom scoping
const config: VendureConfig = {
    settingsStoreFields: {
        analytics: [
            {
                name: 'quarterlyTargets',
                scope: customDateBasedScope, // Custom quarterly user-specific scope
                public: false,
                requiresPermission: [Permission.UpdateSettings]
            }
        ]
    }
};
```
