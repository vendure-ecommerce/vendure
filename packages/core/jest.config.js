module.exports = {
    coverageDirectory: 'coverage',
    moduleFileExtensions: ['js', 'json', 'ts'],
    preset: 'ts-jest',
    rootDir: __dirname,
    roots: ['<rootDir>/src', '<rootDir>/mock-data'],
    transform: {
        '^.+\\.(t|j)s$': 'ts-jest',
    },
    globals: {
        'ts-jest': {
            tsconfig: {
                allowJs: true,
            },
        },
    },
    testEnvironment: 'node',
};
