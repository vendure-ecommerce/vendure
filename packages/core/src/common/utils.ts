import { AssetType } from '@vendure/common/lib/generated-types';
import { ID } from '@vendure/common/lib/shared-types';
import { Observable, Observer } from 'rxjs';

/**
 * Takes a predicate function and returns a negated version.
 */
export function not(predicate: (...args: any[]) => boolean) {
    return (...args: any[]) => !predicate(...args);
}

/**
 * Returns a predicate function which returns true if the item is found in the set,
 * as determined by a === equality check on the given compareBy property.
 */
export function foundIn<T>(set: T[], compareBy: keyof T) {
    return (item: T) => set.some(t => t[compareBy] === item[compareBy]);
}

/**
 * Identity function which asserts to the type system that a promise which can resolve to T or undefined
 * does in fact resolve to T.
 * Used when performing a "find" operation on an entity which we are sure exists, as in the case that we
 * just successfully created or updated it.
 */
export function assertFound<T>(promise: Promise<T | undefined>): Promise<T> {
    return promise as Promise<T>;
}

/**
 * Compare ID values for equality, taking into account the fact that they may not be of matching types
 * (string or number).
 */
export function idsAreEqual(id1?: ID, id2?: ID): boolean {
    if (id1 === undefined || id2 === undefined) {
        return false;
    }
    return id1.toString() === id2.toString();
}

/**
 * Returns the AssetType based on the mime type.
 */
export function getAssetType(mimeType: string): AssetType {
    const type = mimeType.split('/')[0];
    switch (type) {
        case 'image':
            return AssetType.IMAGE;
        case 'video':
            return AssetType.VIDEO;
        default:
            return AssetType.BINARY;
    }
}

/**
 * A simple normalization for email addresses. Lowercases the whole address,
 * even though technically the local part (before the '@') is case-sensitive
 * per the spec. In practice, however, it seems safe to treat emails as
 * case-insensitive to allow for users who might vary the usage of
 * upper/lower case. See more discussion here: https://ux.stackexchange.com/a/16849
 */
export function normalizeEmailAddress(input: string): string {
    return input.trim().toLowerCase();
}

/**
 * Converts a value that may be wrapped into a Promise or Observable into a Promise-wrapped
 * value.
 */
export async function awaitPromiseOrObservable<T>(value: T | Promise<T> | Observable<T>): Promise<T> {
    let result = await value;
    if (result instanceof Observable) {
        result = await result.toPromise();
    }
    return result;
}

/**
 * @description
 * Returns an observable which executes the given async work function and completes with
 * the returned value. This is useful in methods which need to return
 * an Observable but also want to work with async (Promise-returning) code.
 *
 * @example
 * ```TypeScript
 * \@Controller()
 * export class MyWorkerController {
 *
 *     \@MessagePattern('test')
 *     handleTest() {
 *         return asyncObservable(async observer => {
 *             const value = await this.connection.fetchSomething();
 *             return value;
 *         });
 *     }
 * }
 * ```
 */
export function asyncObservable<T>(work: (observer: Observer<T>) => Promise<T | void>): Observable<T> {
    return new Observable<T>(subscriber => {
        (async () => {
            try {
                const result = await work(subscriber);
                if (result) {
                    subscriber.next(result);
                }
                subscriber.complete();
            } catch (e) {
                subscriber.error(e);
            }
        })();
    });
}
