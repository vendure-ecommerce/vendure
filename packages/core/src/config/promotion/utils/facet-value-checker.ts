import { ID } from '@vendure/common/lib/shared-types';
import { unique } from '@vendure/common/lib/unique';
import { RequestContext } from '../../../api';

import { TtlCache } from '../../../common/ttl-cache';
import { idsAreEqual } from '../../../common/utils';
import { TransactionalConnection } from '../../../connection/transactional-connection';
import { OrderLine } from '../../../entity/order-line/order-line.entity';
import { ProductVariant } from '../../../entity/product-variant/product-variant.entity';

/**
 * @description
 * The FacetValueChecker is a helper class used to determine whether a given OrderLine consists
 * of ProductVariants containing the given FacetValues.
 *
 * @example
 * ```TypeScript
 * import { FacetValueChecker, LanguageCode, PromotionCondition, TransactionalConnection } from '\@vendure/core';
 *
 * let facetValueChecker: FacetValueChecker;
 *
 * export const hasFacetValues = new PromotionCondition({
 *   code: 'at_least_n_with_facets',
 *   description: [
 *     { languageCode: LanguageCode.en, value: 'Buy at least { minimum } products with the given facets' },
 *   ],
 *   args: {
 *     minimum: { type: 'int' },
 *     facets: { type: 'ID', list: true, ui: { component: 'facet-value-form-input' } },
 *   },
 *   init(injector) {
 *     facetValueChecker = new FacetValueChecker(injector.get(TransactionalConnection));
 *   },
 *   // tslint:disable-next-line:no-shadowed-variable
 *   async check(ctx, order, args) {
 *     let matches = 0;
 *     for (const line of order.lines) {
 *       if (await facetValueChecker.hasFacetValues(line, args.facets)) {
 *           matches += line.quantity;
 *       }
 *     }
 *     return args.minimum <= matches;
 *   },
 * });
 * ```
 *
 * @docsCategory Promotions
 */
export class FacetValueChecker {
    private variantCache = new TtlCache<ID, ProductVariant>({ ttl: 5000 });

    constructor(private connection: TransactionalConnection) {}
    /**
     * @description
     * Checks a given {@link OrderLine} against the facetValueIds and returns
     * `true` if the associated {@link ProductVariant} & {@link Product} together
     * have *all* the specified {@link FacetValue}s.
     */
    async hasFacetValues(orderLine: OrderLine, facetValueIds: ID[], ctx?: RequestContext): Promise<boolean> {
        let variant = this.variantCache.get(orderLine.productVariant.id);
        if (!variant) {
            variant = await this.connection
                .getRepository(ctx, ProductVariant)
                .findOne(orderLine.productVariant.id, {
                    relations: ['product', 'product.facetValues', 'facetValues'],
                });
            if (!variant) {
                return false;
            }
            this.variantCache.set(variant.id, variant);
        }
        const allFacetValues = unique([...variant.facetValues, ...variant.product.facetValues], 'id');
        return facetValueIds.reduce(
            (result, id) => result && !!allFacetValues.find(fv => idsAreEqual(fv.id, id)),
            true as boolean,
        );
    }
}
