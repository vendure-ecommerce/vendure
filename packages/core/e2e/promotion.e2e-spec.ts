import { pick } from '@vendure/common/lib/pick';
import { PromotionAction, PromotionCondition, PromotionOrderAction } from '@vendure/core';
import {
    createErrorResultGuard,
    createTestEnvironment,
    E2E_DEFAULT_CHANNEL_TOKEN,
    ErrorResultGuard,
} from '@vendure/testing';
import gql from 'graphql-tag';
import path from 'path';

import { initialData } from '../../../e2e-common/e2e-initial-data';
import { testConfig, TEST_SETUP_TIMEOUT_MS } from '../../../e2e-common/test-config';

import { PROMOTION_FRAGMENT } from './graphql/fragments';
import {
    AssignPromotionToChannel,
    ChannelFragment,
    CreateChannel,
    CreatePromotion,
    CurrencyCode,
    DeletePromotion,
    DeletionResult,
    ErrorCode,
    GetAdjustmentOperations,
    GetPromotion,
    GetPromotionList,
    LanguageCode,
    Promotion,
    PromotionFragment,
    RemovePromotionFromChannel,
    UpdatePromotion,
} from './graphql/generated-e2e-admin-types';
import {
    ASSIGN_PROMOTIONS_TO_CHANNEL,
    CREATE_CHANNEL,
    CREATE_PROMOTION,
    REMOVE_PROMOTIONS_FROM_CHANNEL,
} from './graphql/shared-definitions';
import { assertThrowsWithMessage } from './utils/assert-throws-with-message';

// tslint:disable:no-non-null-assertion

describe('Promotion resolver', () => {
    const promoCondition = generateTestCondition('promo_condition');
    const promoCondition2 = generateTestCondition('promo_condition2');
    const promoAction = generateTestAction('promo_action');

    const { server, adminClient, shopClient } = createTestEnvironment({
        ...testConfig,
        promotionOptions: {
            promotionConditions: [promoCondition, promoCondition2],
            promotionActions: [promoAction],
        },
    });

    const snapshotProps: Array<keyof Promotion.Fragment> = [
        'name',
        'actions',
        'conditions',
        'enabled',
        'couponCode',
        'startsAt',
        'endsAt',
    ];
    let promotion: Promotion.Fragment;

    const promotionGuard: ErrorResultGuard<PromotionFragment> = createErrorResultGuard(
        input => !!input.couponCode,
    );

    beforeAll(async () => {
        await server.init({
            initialData,
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-minimal.csv'),
            customerCount: 1,
        });
        await adminClient.asSuperAdmin();
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    it('createPromotion', async () => {
        const { createPromotion } = await adminClient.query<
            CreatePromotion.Mutation,
            CreatePromotion.Variables
        >(CREATE_PROMOTION, {
            input: {
                name: 'test promotion',
                enabled: true,
                couponCode: 'TEST123',
                startsAt: new Date('2019-10-30T00:00:00.000Z'),
                endsAt: new Date('2019-12-01T00:00:00.000Z'),
                conditions: [
                    {
                        code: promoCondition.code,
                        arguments: [{ name: 'arg', value: '500' }],
                    },
                ],
                actions: [
                    {
                        code: promoAction.code,
                        arguments: [
                            {
                                name: 'facetValueIds',
                                value: '["T_1"]',
                            },
                        ],
                    },
                ],
            },
        });
        promotionGuard.assertSuccess(createPromotion);

        promotion = createPromotion;
        expect(pick(promotion, snapshotProps)).toMatchSnapshot();
    });

    it('createPromotion return error result with empty conditions and no couponCode', async () => {
        const { createPromotion } = await adminClient.query<
            CreatePromotion.Mutation,
            CreatePromotion.Variables
        >(CREATE_PROMOTION, {
            input: {
                name: 'bad promotion',
                enabled: true,
                conditions: [],
                actions: [
                    {
                        code: promoAction.code,
                        arguments: [
                            {
                                name: 'facetValueIds',
                                value: '["T_1"]',
                            },
                        ],
                    },
                ],
            },
        });
        promotionGuard.assertErrorResult(createPromotion);

        expect(createPromotion.message).toBe(
            'A Promotion must have either at least one condition or a coupon code set',
        );
        expect(createPromotion.errorCode).toBe(ErrorCode.MISSING_CONDITIONS_ERROR);
    });

    it('updatePromotion', async () => {
        const { updatePromotion } = await adminClient.query<
            UpdatePromotion.Mutation,
            UpdatePromotion.Variables
        >(UPDATE_PROMOTION, {
            input: {
                id: promotion.id,
                couponCode: 'TEST1235',
                startsAt: new Date('2019-05-30T22:00:00.000Z'),
                endsAt: new Date('2019-06-01T22:00:00.000Z'),
                conditions: [
                    {
                        code: promoCondition.code,
                        arguments: [{ name: 'arg', value: '90' }],
                    },
                    {
                        code: promoCondition2.code,
                        arguments: [{ name: 'arg', value: '10' }],
                    },
                ],
            },
        });
        promotionGuard.assertSuccess(updatePromotion);

        expect(pick(updatePromotion, snapshotProps)).toMatchSnapshot();
    });

    it('updatePromotion return error result with empty conditions and no couponCode', async () => {
        const { updatePromotion } = await adminClient.query<
            UpdatePromotion.Mutation,
            UpdatePromotion.Variables
        >(UPDATE_PROMOTION, {
            input: {
                id: promotion.id,
                couponCode: '',
                conditions: [],
            },
        });
        promotionGuard.assertErrorResult(updatePromotion);

        expect(updatePromotion.message).toBe(
            'A Promotion must have either at least one condition or a coupon code set',
        );
        expect(updatePromotion.errorCode).toBe(ErrorCode.MISSING_CONDITIONS_ERROR);
    });

    it('promotion', async () => {
        const result = await adminClient.query<GetPromotion.Query, GetPromotion.Variables>(GET_PROMOTION, {
            id: promotion.id,
        });

        expect(result.promotion!.name).toBe(promotion.name);
    });

    it('promotions', async () => {
        const result = await adminClient.query<GetPromotionList.Query, GetPromotionList.Variables>(
            GET_PROMOTION_LIST,
            {},
        );

        expect(result.promotions.totalItems).toBe(1);
        expect(result.promotions.items[0].name).toBe('test promotion');
    });

    it('adjustmentOperations', async () => {
        const result = await adminClient.query<
            GetAdjustmentOperations.Query,
            GetAdjustmentOperations.Variables
        >(GET_ADJUSTMENT_OPERATIONS);

        expect(result.promotionActions).toMatchSnapshot();
        expect(result.promotionConditions).toMatchSnapshot();
    });

    describe('channels', () => {
        const SECOND_CHANNEL_TOKEN = 'SECOND_CHANNEL_TOKEN';
        let secondChannel: ChannelFragment;
        beforeAll(async () => {
            const { createChannel } = await adminClient.query<
                CreateChannel.Mutation,
                CreateChannel.Variables
            >(CREATE_CHANNEL, {
                input: {
                    code: 'second-channel',
                    token: SECOND_CHANNEL_TOKEN,
                    defaultLanguageCode: LanguageCode.en,
                    pricesIncludeTax: true,
                    currencyCode: CurrencyCode.EUR,
                    defaultTaxZoneId: 'T_1',
                    defaultShippingZoneId: 'T_2',
                },
            });
            secondChannel = createChannel as any;
        });

        it('does not list Promotions not in active channel', async () => {
            adminClient.setChannelToken(SECOND_CHANNEL_TOKEN);
            const { promotions } = await adminClient.query<GetPromotionList.Query>(GET_PROMOTION_LIST);

            expect(promotions.totalItems).toBe(0);
            expect(promotions.items).toEqual([]);
        });

        it('does not return Promotion not in active channel', async () => {
            adminClient.setChannelToken(SECOND_CHANNEL_TOKEN);
            const { promotion: result } = await adminClient.query<GetPromotion.Query, GetPromotion.Variables>(
                GET_PROMOTION,
                {
                    id: promotion.id,
                },
            );

            expect(result).toBeNull();
        });

        it('assignPromotionsToChannel', async () => {
            adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
            const { assignPromotionsToChannel } = await adminClient.query<
                AssignPromotionToChannel.Mutation,
                AssignPromotionToChannel.Variables
            >(ASSIGN_PROMOTIONS_TO_CHANNEL, {
                input: {
                    channelId: secondChannel.id,
                    promotionIds: [promotion.id],
                },
            });

            expect(assignPromotionsToChannel).toEqual([{ id: promotion.id, name: promotion.name }]);

            adminClient.setChannelToken(SECOND_CHANNEL_TOKEN);
            const { promotion: result } = await adminClient.query<GetPromotion.Query, GetPromotion.Variables>(
                GET_PROMOTION,
                {
                    id: promotion.id,
                },
            );
            expect(result?.id).toBe(promotion.id);

            const { promotions } = await adminClient.query<GetPromotionList.Query>(GET_PROMOTION_LIST);
            expect(promotions.totalItems).toBe(1);
            expect(promotions.items.map(pick(['id']))).toEqual([{ id: promotion.id }]);
        });

        it('removePromotionsFromChannel', async () => {
            adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
            const { removePromotionsFromChannel } = await adminClient.query<
                RemovePromotionFromChannel.Mutation,
                RemovePromotionFromChannel.Variables
            >(REMOVE_PROMOTIONS_FROM_CHANNEL, {
                input: {
                    channelId: secondChannel.id,
                    promotionIds: [promotion.id],
                },
            });

            expect(removePromotionsFromChannel).toEqual([{ id: promotion.id, name: promotion.name }]);

            adminClient.setChannelToken(SECOND_CHANNEL_TOKEN);
            const { promotion: result } = await adminClient.query<GetPromotion.Query, GetPromotion.Variables>(
                GET_PROMOTION,
                {
                    id: promotion.id,
                },
            );
            expect(result).toBeNull();

            const { promotions } = await adminClient.query<GetPromotionList.Query>(GET_PROMOTION_LIST);
            expect(promotions.totalItems).toBe(0);
        });
    });

    describe('deletion', () => {
        let allPromotions: GetPromotionList.Items[];
        let promotionToDelete: GetPromotionList.Items;

        beforeAll(async () => {
            adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
            const result = await adminClient.query<GetPromotionList.Query>(GET_PROMOTION_LIST);
            allPromotions = result.promotions.items;
        });

        it('deletes a promotion', async () => {
            promotionToDelete = allPromotions[0];
            const result = await adminClient.query<DeletePromotion.Mutation, DeletePromotion.Variables>(
                DELETE_PROMOTION,
                { id: promotionToDelete.id },
            );

            expect(result.deletePromotion).toEqual({ result: DeletionResult.DELETED });
        });

        it('cannot get a deleted promotion', async () => {
            const result = await adminClient.query<GetPromotion.Query, GetPromotion.Variables>(
                GET_PROMOTION,
                {
                    id: promotionToDelete.id,
                },
            );

            expect(result.promotion).toBe(null);
        });

        it('deleted promotion omitted from list', async () => {
            const result = await adminClient.query<GetPromotionList.Query>(GET_PROMOTION_LIST);

            expect(result.promotions.items.length).toBe(allPromotions.length - 1);
            expect(result.promotions.items.map(c => c.id).includes(promotionToDelete.id)).toBe(false);
        });

        it(
            'updatePromotion throws for deleted promotion',
            assertThrowsWithMessage(
                () =>
                    adminClient.query<UpdatePromotion.Mutation, UpdatePromotion.Variables>(UPDATE_PROMOTION, {
                        input: {
                            id: promotionToDelete.id,
                            enabled: false,
                        },
                    }),
                `No Promotion with the id '1' could be found`,
            ),
        );
    });
});

function generateTestCondition(code: string): PromotionCondition<any> {
    return new PromotionCondition({
        code,
        description: [{ languageCode: LanguageCode.en, value: `description for ${code}` }],
        args: { arg: { type: 'int' } },
        check: (order, args) => true,
    });
}

function generateTestAction(code: string): PromotionAction<any> {
    return new PromotionOrderAction({
        code,
        description: [{ languageCode: LanguageCode.en, value: `description for ${code}` }],
        args: { facetValueIds: { type: 'ID', list: true } },
        execute: (order, args) => {
            return 42;
        },
    });
}

const DELETE_PROMOTION = gql`
    mutation DeletePromotion($id: ID!) {
        deletePromotion(id: $id) {
            result
        }
    }
`;

export const GET_PROMOTION_LIST = gql`
    query GetPromotionList($options: PromotionListOptions) {
        promotions(options: $options) {
            items {
                ...Promotion
            }
            totalItems
        }
    }
    ${PROMOTION_FRAGMENT}
`;

export const GET_PROMOTION = gql`
    query GetPromotion($id: ID!) {
        promotion(id: $id) {
            ...Promotion
        }
    }
    ${PROMOTION_FRAGMENT}
`;

export const UPDATE_PROMOTION = gql`
    mutation UpdatePromotion($input: UpdatePromotionInput!) {
        updatePromotion(input: $input) {
            ...Promotion
            ... on ErrorResult {
                errorCode
                message
            }
        }
    }
    ${PROMOTION_FRAGMENT}
`;

export const CONFIGURABLE_DEF_FRAGMENT = gql`
    fragment ConfigurableOperationDef on ConfigurableOperationDefinition {
        args {
            name
            type
            ui
        }
        code
        description
    }
`;

export const GET_ADJUSTMENT_OPERATIONS = gql`
    query GetAdjustmentOperations {
        promotionActions {
            ...ConfigurableOperationDef
        }
        promotionConditions {
            ...ConfigurableOperationDef
        }
    }
    ${CONFIGURABLE_DEF_FRAGMENT}
`;
