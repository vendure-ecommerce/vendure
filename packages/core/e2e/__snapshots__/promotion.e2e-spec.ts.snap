// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Promotion resolver > adjustmentOperations 1`] = `
[
  {
    "args": [
      {
        "name": "facetValueIds",
        "type": "ID",
        "ui": null,
      },
    ],
    "code": "promo_action",
    "description": "description for promo_action",
  },
]
`;

exports[`Promotion resolver > adjustmentOperations 2`] = `
[
  {
    "args": [
      {
        "name": "arg",
        "type": "int",
        "ui": null,
      },
    ],
    "code": "promo_condition",
    "description": "description for promo_condition",
  },
  {
    "args": [
      {
        "name": "arg",
        "type": "int",
        "ui": null,
      },
    ],
    "code": "promo_condition2",
    "description": "description for promo_condition2",
  },
]
`;

exports[`Promotion resolver > createPromotion 1`] = `
{
  "actions": [
    {
      "args": [
        {
          "name": "facetValueIds",
          "value": "[\\"T_1\\"]",
        },
      ],
      "code": "promo_action",
    },
  ],
  "conditions": [
    {
      "args": [
        {
          "name": "arg",
          "value": "500",
        },
      ],
      "code": "promo_condition",
    },
  ],
  "couponCode": "TEST123",
  "enabled": true,
  "endsAt": "2019-12-01T00:00:00.000Z",
  "name": "test promotion",
  "startsAt": "2019-10-30T00:00:00.000Z",
}
`;

exports[`Promotion resolver > updatePromotion 1`] = `
{
  "actions": [
    {
      "args": [
        {
          "name": "facetValueIds",
          "value": "[\\"T_1\\"]",
        },
      ],
      "code": "promo_action",
    },
  ],
  "conditions": [
    {
      "args": [
        {
          "name": "arg",
          "value": "90",
        },
      ],
      "code": "promo_condition",
    },
    {
      "args": [
        {
          "name": "arg",
          "value": "10",
        },
      ],
      "code": "promo_condition2",
    },
  ],
  "couponCode": "TEST1235",
  "enabled": true,
  "endsAt": "2019-06-01T22:00:00.000Z",
  "name": "test promotion",
  "startsAt": "2019-05-30T22:00:00.000Z",
}
`;
